rfc-172-graphql-for-govuk.md

```diff
@@ -0,0 +1,509 @@
+|        Index Cond: ((document_id = document.id) AND ((state)::text = 'published'::text))                                                                                                                           |
+|Planning Time: 1.115 ms                                                                                                                                                                                             |
+|Execution Time: 0.318 ms                                                                                                                                                                                            |
++--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+```
+
+Which can be [explored visually in PG Explain][pg-explain-query-plan]
+
+![PG Explain Visualizer showing the execution plan above](rfc-172/pgexplain-query-plan-visualisation.png)
+
+</details>
+
+Still, even with dataloader batching queries and individual queries making good use of indexes, we're asking the database to do at least around 10x more work than the current content-store database has to do, and that will have some performance or infrastructure cost implications. We will need to carefully consider how this solution performs at scale to evaluate these.
+
+Unlike currently, more complicated pages (with lots of links) will be more expensive to render than simple pages. The most complicated page by far on GOV.UK currently is the ministers index page, and in the first prototype GraphQL API this takes between ~500ms and ~1000ms to generate an API response - real world performance may differ, but it will be important to evaluate the performance of the more complicated pages.
+
+In terms of page load times as seen by an end user, there are additional factors that mitigate the impact of API performance. Mainly, the impact of caching - the ~90% of requests which we serve from Fastly's cache won't be affected at all by API performance, while the 10% of cache misses will. We could consider further tuning of our cache behaviour (for example using [stale-while-revalidate][stale-while-revalidate]) to further reduce performance impact.
+
+### Frontend applications will need to make two API calls instead of one
+
+Currently, frontend applications only need to make a single request to content store, and they can then switch on the document type in the response to determine which view to render.
+
+The value of the GraphQL approach is that the frontend can specify the fields it needs for the document type it's rendering. For this to work, the frontend needs to know which document type a particular page has before it can make the query to get the page data.
```

Comment:
Could this be mitigated by some kind of "standardised fields" option? I can't remember if GraphQL has a way of expressing this. I'm envisioning something like "Let me know what type of content this is, and also return the usual fields returned for this sort of content". I guess a little bit like a prepared statement in SQL. Then the usual fields are the ones required to render that type of page, and we only need to do one call. It does mean that there's a reduction in flexibility (if we want to add additional fields and keep the single call we have to manually tell the API that new fields are needed), but as an option if we really need the speed of just a single call, it's something.

Reply:
Yes, I think you can achieve this [with interfaces / unions and fragments](https://graphql.com/learn/interfaces-and-unions/#fragments). So you could have one giant query like:

```graphql
query {
  edition(basePath: $basePath) {
    publicLastUpdatedAt
    # ... Other fields that are present on all editions
    
    ... on Guide {
      bodyHtml
      # ...
    }
    ... on Manual {
       parts
       # ...
    }
    ... on MinistersIndex {
       cabinetMinisters { }
       # ...
    }
    # ... ~180 other content types
  }
}
```

This would be a big request, but we could store it as a [persisted query](https://graphql-ruby.org/operation_store/overview.html).